### PR TITLE
Fix/issue 1512

### PR DESCRIPTION
**Pull request by iparask**
This pull request address issue #1512 based on the discussion on that ticket.

### CODE REVIEW

File: examples/05_unit_input_data.py [hunk 1]
```diff
@@ -35,7 +35,7 @@
 
     # Create a new session. No need to try/except this: if session creation
     # fails, there is not much we can do anyways...
-    session = rp.Session()
+    session = rp.Session(uid='TarTest')
```
Comment (andre-merzky) [inline examples/05_unit_input_data.py]: Please don't hardcode session names in the examples - this will fail on the second execution.

File: examples/05_unit_input_data.py [hunk 2]
```diff
@@ -92,12 +94,15 @@
...
           # this is a shortcut for:
-          # cud.input_staging  = {'source': 'client:///input.dat', 
-          #                       'target': 'unit:///input.dat',
-          #                       'action': rp.Transfer}
+            cud.input_staging  = [{'source': 'client:///input.dat', 
+                                   'target': 'unit:///input.dat',
+                                   'action': rp.TARBALL},
+                                  {'source': 'client:///input2.dat', 
+                                   'target': 'unit:///input2.dat',
+                                   'action': rp.TARBALL}]
```
Comment (andre-merzky) [inline examples/05_unit_input_data.py]: Wouldn't it make sense to add an additional example?  I would think we should keep the original example on the canonical staging (otherwise we'll need to explain what  TARBALL means...)

File: src/radical/pilot/agent/staging_input/default.py [hunk 3]
```diff
@@ -103,7 +104,7 @@ def _handle_unit(self, unit, actionables):
         # NOTE: see documentation of cu['sandbox'] semantics in the ComputeUnit
         #       class definition.
         sandbox = unit['unit_sandbox']
-
+
```
Comment (andre-merzky) [inline src/radical/pilot/agent/staging_input/default.py]: why?  :-)

File: src/radical/pilot/agent/staging_input/default.py [hunk 4]
```diff
@@ -199,7 +200,13 @@ def _handle_unit(self, unit, actionables):
                     self._log.error('no transfer for %s -> %s', src, tgt)
                     self._prof.prof('staging_in_fail', uid=uid, msg=did)
                     raise NotImplementedError('unsupported transfer %s' % src)
-
+            elif action == rpc.TARBALL:
+                # If somethig was staged via the tarball method, the tarball is
+                # extracted and then removed from the unit folder.
+                tar = tarfile.open(tgtdir+'/'+uid+'.tar')
```
Comment (andre-merzky) [inline src/radical/pilot/agent/staging_input/default.py]: please use spaces around `+` (PEP.8).

Comment (andre-merzky) [inline src/radical/pilot/agent/staging_input/default.py]: Per the `if` above, you can get `client://` type URLs here - do those need expansion?

File: src/radical/pilot/agent/staging_input/default.py [hunk 5]
```diff
@@ -199,7 +200,13 @@ def _handle_unit(self, unit, actionables):
...
+                tar.extractall(path=tgtdir)
+                tar.close()
+                os.remove(tgtdir+'/'+uid+'.tar')
```
Comment (andre-merzky) [inline src/radical/pilot/agent/staging_input/default.py]: You may want to keep the tarballs if `RADICAL_DEBUG` is set?

File: src/radical/pilot/umgr/staging_input/default.py [hunk 6]
```diff
@@ -325,26 +325,58 @@ def _handle_unit(self, unit, actionables):
...
+                # and append the directive to the removables.
 
-            self._prof.prof('staging_in_start', uid=uid, msg=did)
+                if not os.path.isfile(uid+'.tar'):
+                    tar_file = tarfile.open(uid+'.tar','w')
+                tar_file.add(src.split('client:///')[1],arcname=tgt.split('unit:///')[1])
+                removables.append(sd)
+                sid = sd.get('uid')
+            elif action == rpc.TRANSFER:
+                self._prof.prof('staging_in_start', uid=uid, msg=did)
```
Comment (andre-merzky) [inline src/radical/pilot/umgr/staging_input/default.py]: You moved the profile entry into the `elif` - now the event is  missing for the tarballs.

File: src/radical/pilot/umgr/staging_input/default.py [hunk 7]
```diff
@@ -325,26 +325,58 @@ def _handle_unit(self, unit, actionables):
...
             did    = sd['uid']
             src    = sd['source']
             tgt    = sd['target']
+            if action == rpc.TARBALL:
+                # If the staging method is tarball, check initially if the tar 
+                # file exists. If not create it. Add each file in the tarball
+                # and append the directive to the removables.
 
-            self._prof.prof('staging_in_start', uid=uid, msg=did)
+                if not os.path.isfile(uid+'.tar'):
```
Comment (andre-merzky) [inline src/radical/pilot/umgr/staging_input/default.py]: please use spaces around `+`

File: src/radical/pilot/umgr/staging_input/default.py [hunk 8]
```diff
@@ -325,26 +325,58 @@ def _handle_unit(self, unit, actionables):
...
+                    tar_file = tarfile.open(uid+'.tar','w')
+                tar_file.add(src.split('client:///')[1],arcname=tgt.split('unit:///')[1])
```
Comment (andre-merzky) [inline src/radical/pilot/umgr/staging_input/default.py]: Are you *sure* that the src and tgt specs match that pattern here?  Maybe use URL parsing and then `url.path`?  This differs from how the `elif` branch handles the URLs.

File: src/radical/pilot/umgr/staging_input/default.py [hunk 9]
```diff
@@ -325,26 +325,58 @@ def _handle_unit(self, unit, actionables):
...
+                removables.append(sd)
+                sid = sd.get('uid')
```
Comment (andre-merzky) [inline src/radical/pilot/umgr/staging_input/default.py]: this is the same as `did` above?

File: src/radical/pilot/umgr/staging_input/default.py [hunk 10]
```diff
@@ -325,26 +325,58 @@ def _handle_unit(self, unit, actionables):
...
 
-            if rpc.CREATE_PARENTS in flags:
-                copy_flags = rs.filesystem.CREATE_PARENTS
-            else:
-                copy_flags = 0
+                saga_dir.copy(src, tgt, flags=flags)
+                self._prof.prof('staging_in_stop', uid=uid, msg=did)
+
+
+        if os.path.isfile(uid+'.tar'):
```
Comment (andre-merzky) [inline src/radical/pilot/umgr/staging_input/default.py]: Wouldn't it make sense to check if the `tar_file` object exists instead of hitting the file system?

File: src/radical/pilot/umgr/staging_input/default.py [hunk 11]
```diff
@@ -325,26 +325,58 @@ def _handle_unit(self, unit, actionables):
...
+            # if a tarball exists close it and do the transfer. After it is done
+            # delete the tarball. Remove any directive that was use to create the
+            # tarball and add a single one that communicates to the agent that
+            # this type of transfer was executed.
+            tar_file.close()
+
+            self._prof.prof('staging_in_start', uid=uid, msg=did)
```
Comment (andre-merzky) [inline src/radical/pilot/umgr/staging_input/default.py]: Ah, here it is.  So the transfer ID is then the did of the last staging directive.  Hmm, Lets discuss this, I am not sure what analytics will think if some SDs do not have any profile events at all.  The files get transferred after all...

File: src/radical/pilot/umgr/staging_input/default.py [hunk 12]
```diff
@@ -325,26 +325,58 @@ def _handle_unit(self, unit, actionables):
...
+
+            src = complete_url(uid+'.tar', src_context, self._log)
+            tgt = complete_url(uid+'.tar', tgt_context, self._log)
 
-            saga_dir.copy(src, tgt, flags=copy_flags)
+            saga_dir.copy(src, tgt, flags=rs.filesystem.CREATE_PARENTS)
```
Comment (andre-merzky) [inline src/radical/pilot/umgr/staging_input/default.py]: You hardcode that flag here - that is different from the TRANSFER case.

File: src/radical/pilot/umgr/staging_input/default.py [hunk 13]
```diff
@@ -325,26 +325,58 @@ def _handle_unit(self, unit, actionables):
...
             self._prof.prof('staging_in_stop', uid=uid, msg=did)
+            for sd in removables:
+                unit['description']['input_staging'].remove(sd)
+            unit['description']['input_staging'].append({'uid': sid, 
+                                                         'priority': 0, 
+                                                         'source': 'client:///'+uid+'.tar', 
+                                                         'flags': ['CreateParents', 'SkipFailed'],
```
Comment (andre-merzky) [inline src/radical/pilot/umgr/staging_input/default.py]: please use `DEFAULT_FLAGS`

File: src/radical/pilot/umgr/staging_input/default.py [hunk 14]
```diff
@@ -325,26 +325,58 @@ def _handle_unit(self, unit, actionables):
...
-            saga_dir.copy(src, tgt, flags=copy_flags)
+            saga_dir.copy(src, tgt, flags=rs.filesystem.CREATE_PARENTS)
             self._prof.prof('staging_in_stop', uid=uid, msg=did)
+            for sd in removables:
+                unit['description']['input_staging'].remove(sd)
+            unit['description']['input_staging'].append({'uid': sid, 
+                                                         'priority': 0, 
+                                                         'source': 'client:///'+uid+'.tar', 
+                                                         'flags': ['CreateParents', 'SkipFailed'], 
+                                                         'action': 'Tarball',
```
Comment (andre-merzky) [inline src/radical/pilot/umgr/staging_input/default.py]: please use the constant.

File: tests/issue_1512/run_test.sh [hunk 15]
```diff
@@ -0,0 +1,44 @@
+#!/bin/bash
+
+for jj in {1,2,4,8,16,32,64,128,256,512,1024,2048}; do
```
Comment (andre-merzky) [inline tests/issue_1512/run_test.sh]: Why the array syntax?  This works just as well and is simpler:
```
$ for i in 1 2 3; do echo $i; done
1
2
3
```

Doesn't matter really - but in general you may want to use POSIX where bash specific features are not needed, to keep scripts portable.

File: tests/issue_1512/TarCpUntar.py [hunk 16]
```diff
@@ -0,0 +1,27 @@
+import os
```
Comment (andre-merzky) [inline tests/issue_1512/TarCpUntar.py]: Please apply the coding guidelines also to tests :-)
More importantly,  how is that code related to testing the #1512?   It looks like you test the tarfile module - that is not our business...

File: tests/issue_1512/randomFiles.py [hunk 17]
```diff
@@ -0,0 +1,28 @@
+import string
```
Comment (andre-merzky) [inline tests/issue_1512/randomFiles.py]: FYI, this is a much quicker and simpler way to create files of a certain size in a shell script:
```
$ time head -c $((1024*1024)) /dev/urandom > /tmp/t.dat
r:0m0.012s  u:0m0.001s  s:0m0.011s
```

File: tests/issue_1512/run_test.sh [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+    python randomFiles.py $jj
+    echo '----------Number of files: '$jj'----------'
+    echo '---------- Half Meg ----------'
+    echo 'Time'>run_half_$jj.csv
+    for ii in {1..10}; do
+            python transfer_files.py halfMeg wrangler.tacc.utexas.edu <path> >>run_half_$jj.csv 2>>log.log
```
Comment (andre-merzky) [inline tests/issue_1512/run_test.sh]: I am missing something - what is `<path>` here?  `> >>` looks strange to me, too.  Does this run?

File: tests/issue_1512/transfer_files.py [hunk 19]
```diff
@@ -0,0 +1,73 @@
+import os
```
Comment (andre-merzky) [inline tests/issue_1512/transfer_files.py]: please apply coding guidelines.

File: tests/issue_1512/transfer_files.py [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+if __name__ == '__main__':
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument('srcfolder',help='Folder which contains the files to be transfered')
+    parser.add_argument('resource', help="Target Resource that support GSISFTP")
+    parser.add_argument('tgtfolder', help="Target folder. Should be the whole path.")
+    parser.add_argument('--tar',help='Transfer without tarball or with',default=False,action="store_true")
+    args = parser.parse_args()
+    
+    TansferFiles = glob.glob(args.srcfolder+'/*')
```
Comment (andre-merzky) [inline tests/issue_1512/transfer_files.py]: typo (T*r*ansfer)

File: src/radical/pilot/umgr/staging_input/default.py (hunk 6 above)
Comment (iparask) [reply src/radical/pilot/umgr/staging_input/default.py]: Yes and also I need to create a new id for this staging

File: src/radical/pilot/agent/staging_input/default.py (hunk 3 above)
Comment (iparask) [reply src/radical/pilot/agent/staging_input/default.py]: I think I had something there and I deleted it.

File: examples/05_unit_input_data.py (hunk 1 above)
Comment (iparask) [reply examples/05_unit_input_data.py]: Done

File: examples/05_unit_input_data.py (hunk 2 above)
Comment (iparask) [reply examples/05_unit_input_data.py]: Done

File: src/radical/pilot/agent/staging_input/default.py (hunk 3 above)
Comment (iparask) [reply src/radical/pilot/agent/staging_input/default.py]: done

File: src/radical/pilot/agent/staging_input/default.py (hunk 4 above)
Comment (iparask) [reply src/radical/pilot/agent/staging_input/default.py]: done

Comment (iparask) [reply src/radical/pilot/agent/staging_input/default.py]: I do not think there is any need to check about `client:///`. The `if` in line 161 makes sure that whatever passes does not have `client://` at the target. Am I right?

I have a feeling that either I am missing something major,

File: src/radical/pilot/agent/staging_input/default.py (hunk 5 above)
Comment (iparask) [reply src/radical/pilot/agent/staging_input/default.py]: Are the files the user transferred enough? I agree it needs a debug message to show what is happening.

**Comment by iparask** [comment]
Hello @andre-merzky , I am not sure how to trigger a new review request. I believe I addressed most of your comments. Also Srinivas has pushed the unit tests. Can you please take another look?

Thank you

File: examples/05_unit_input_data.py [hunk 21]
```diff
@@ -92,12 +92,12 @@
...
-            cud.input_staging  = ['input.dat']
+            #cud.input_staging  = ['input.dat']
             
           # this is a shortcut for:
-          # cud.input_staging  = {'source': 'client:///input.dat', 
-          #                       'target': 'unit:///input.dat',
-          #                       'action': rp.Transfer}
+            cud.input_staging  = [{'source': 'client:///input.dat', 
+                                   'target': 'unit:///input.dat',
+                                   'action': rp.TRANSFER}]
```
Comment (andre-merzky) [inline examples/05_unit_input_data.py]: Wait, why is this a list now?  I am missing something...

File: src/radical/pilot/umgr/staging_input/default.py [hunk 22]
```diff
@@ -325,26 +324,61 @@ def _handle_unit(self, unit, actionables):
...
+                sid = ru.generate_id('sd')
+                self._prof.prof('tarball_create_stop', uid=uid, msg=did)
+            elif action == rpc.TRANSFER:
+                self._prof.prof('staging_in_start', uid=uid, msg=did)
+
+                src = complete_url(src, src_context, self._log)
+                tgt = complete_url(tgt, tgt_context, self._log)
+
+                saga_dir.copy(src, tgt, flags=flags)
+                self._prof.prof('staging_in_stop', uid=uid, msg=did)
```
Comment (andre-merzky) [inline src/radical/pilot/umgr/staging_input/default.py]: This event is duplicated for this uid/did now.

**Comment by andre-merzky** [comment]
Please run at least one remote test on XSEDE.

**Comment by andre-merzky** [comment]
Guys, help me out - how did `tests/test_tarball_umgr.py` actually work when the `saga.fs.Directory.copy` call was `mock`'ed away?  I don't get it...

**Comment by andre-merzky** [comment]
Ah, I do get it now (1 min after last comment :-) : you mask the ENOTDIR exception which happens when passing a URL to the shutil copy...  Man, this took me ages...

Leaving SAGA in place to do the actual thing works.  I would argue that one should only mock those parts which interfere with running the test, really.  Otherwise we don't really test much, do we?  @itomaldonado , any opinion?

**Comment by itomaldonado** [comment]
@andre-merzky I wanted to comment on your last point. I suggested them not testing the saga layer because this test specifically isn't looking at how saga does a transfer. From this module's perspective one the saga function returns successful, the assumption is that the file made it to its destination. We can remove the dependency of Saga (and having to set-up a transfer, ssh setups etc.) by mocking the function call. You can also mock saga to return errors (i.e. raise exceptions) and have unit tests covering the scenarios when that function call fails, but we don't really care about actually testing the transfer because that is not what this test is about.

the assumption is that the saga function was tested separately.

Hopefully that makes sense?
